Objects/typeobject.c

```diff
@@ -60,30 +60,114 @@ class object "PyObject *" "&PyBaseObject_Type"
 #define TYPE_LOCK &PyInterpreterState_Get()->types.mutex
-#define BEGIN_TYPE_LOCK() Py_BEGIN_CRITICAL_SECTION_MUT(TYPE_LOCK)
-#define END_TYPE_LOCK() Py_END_CRITICAL_SECTION()
 
-#define BEGIN_TYPE_DICT_LOCK(d) \
-    Py_BEGIN_CRITICAL_SECTION2_MUT(TYPE_LOCK, &_PyObject_CAST(d)->ob_mutex)
+// Since TYPE_LOCK is used as regular mutex, we must take special care about
+// potential re-entrant code paths.  We use TYPE_LOCK_TID in debug builds to
+// ensure that we are not trying to re-acquire the mutex when it is already
+// held by the current thread.  There are a couple cases when we release the
+// mutex, when we call functions that might re-enter.  If we already hold the
+// lock and try to acquire it again with the same thread, it is a bug in the
+// code.  We could just let it deadlock but having an assert failure gives
+// a more explicit error.
+#define TYPE_LOCK_TID &PyInterpreterState_Get()->types.mutex_tid
+
+// Return true if the world is currently stopped.
+static bool
+types_world_is_stopped(void)
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    return interp->stoptheworld.world_stopped;
+}
+
+// Checks that the type has not yet been revealed (exposed) to other
+// threads.  The _Py_TYPE_REVEALED_FLAG flag is set by type_new() and
+// PyType_FromMetaclass() to indicate that a newly initialized type might be
+// revealed.  We only have ob_flags on 64-bit platforms.
+#if SIZEOF_VOID_P > 4
+#define TYPE_IS_REVEALED(tp) ((((PyObject *)(tp))->ob_flags & _Py_TYPE_REVEALED_FLAG) != 0)
+#else
+#define TYPE_IS_REVEALED(tp) 0
+#endif
 
-#define END_TYPE_DICT_LOCK() Py_END_CRITICAL_SECTION2()
+// Checks if we can safely update type slots or tp_flags.
+#define ASSERT_NEW_OR_STOPPED(tp) \
+    assert(!TYPE_IS_REVEALED(tp) || types_world_is_stopped())
 
-#define ASSERT_TYPE_LOCK_HELD() \
-    _Py_CRITICAL_SECTION_ASSERT_MUTEX_LOCKED(TYPE_LOCK)
+static void
+types_stop_world(void)
+{
+    assert(!types_world_is_stopped());
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    _PyEval_StopTheWorld(interp);
+    assert(types_world_is_stopped());
+}
+
+static void
+types_start_world(void)
+{
+    assert(types_world_is_stopped());
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    _PyEval_StartTheWorld(interp);
+    assert(!types_world_is_stopped());
+}
+
+#ifdef Py_DEBUG
+// Return true if the TYPE_LOCK mutex is held by the current thread.
+static bool
+types_mutex_is_owned(void)
+{
+    PyThread_ident_t tid = PyThread_get_thread_ident_ex();
+    return _Py_atomic_load_ullong_relaxed(TYPE_LOCK_TID) == tid;
+}
+#endif
+
+// Set the TID of the thread currently holding TYPE_LOCK.
+static void
+types_mutex_set_owned(PyThread_ident_t tid)
+{
+    _Py_atomic_store_ullong_relaxed(TYPE_LOCK_TID, tid);
+}
+
+static void
+types_mutex_lock(void)
+{
+    assert(!types_mutex_is_owned());
+    PyMutex_Lock(TYPE_LOCK);
+    types_mutex_set_owned(PyThread_get_thread_ident_ex());
```

Comment:
If this is limited to the free-threading build, I'd be a more comfortable using `Py_ThreadId()` for this and I think it'll be faster.

Reply:
I changed the code but then I removed it after switching back to a critical section.  I think `_Py_ThreadId()` would work fine and be faster, as you say.